### PR TITLE
docker: Migrate the run dialog to use real patterns

### PR DESCRIPTION
**Pull request by stefwalter**
The failures during launch and validation have been rather irritating and in order to debug those we need a solid code base that has predictable behavior.

In particular, there should be no input validation during data entry, no timeouts, side effects of hidden controls, etc. As described in our HIG.

### CODE REVIEW

**Comment by stefwalter** [comment]
I guess this is a duplicate of #4405.

**Comment by dperpeet** [comment]
I will take a look and see if there are any conflicts. Since you've already put in the work, I don't see the harm in simplifying the dialog before the react version lands.

**Comment by dperpeet** [comment]
- When I check "Expose container ports", the dialog adds a line to the links instead.
 - For some reason starting a container seems to take a bit longer - the dialog stays up longer with the spinner, and both buttons (Cancel + Start) are disabled
 - Not sure if this is intentional: when field level errors in the links (e.g. "Duplicate alias") disappear, the box stays large instead of resizing back to its original size as it currently does on master
 - When the field level error "Duplicate alias" isn't applicable anymore (change one, change focus), only one error disappears, not the other (duplicate alias always affects more than one, obviously)

**Comment by dperpeet** [comment]
@stefwalter Depending on how urgent this is, I could put more effort into finishing #4405 instead.

**Comment by stefwalter** [comment]
Blocked on #4726

**Comment by stefwalter** [comment]
>    When I check "Expose container ports", the dialog adds a line to the links instead.

Will fix.

>    For some reason starting a container seems to take a bit longer - the dialog stays up longer with the spinner, and both buttons (Cancel + Start) are disabled

Before the dialog wasn't waiting for the container start. Now it does, as in the rest of Cockpit.

>  Not sure if this is intentional: when field level errors in the links (e.g. "Duplicate alias") disappear, the box stays large instead of resizing back to its original size as it currently does on master

That needs to be fixed.

> When the field level error "Duplicate alias" isn't applicable anymore (change one, change focus), only one error disappears, not the other (duplicate alias always affects more than one, obviously)

This probably needs to be handled specifically.

**Comment by stefwalter** [comment]
> @stefwalter Depending on how urgent this is, I could put more effort into finishing #4405 instead.

Makes sense. It's really the tests that keep falling over and are hard to debug due to the timeouts and races in the code. So if #4405 helps with that, then that's great.

File: pkg/docker/run.js [hunk 1]
```diff
@@ -21,905 +21,699 @@ define([
...
-            function add_row() {
-                render();
+        function remove_row(e) {
+            var parent = $(e.target).closest("form");
+            parent.remove();
+            if ($('#select-claimed-envvars').children().length === 0 ) {
+                $("#claim-envvars").attr("checked", false);
             }
+            /* update run button, this may have removed an error */
+            self.validator("changeFocus", "envvars");
```
Comment (dperpeet) [inline pkg/docker/run.js]: when trying out errors, I get:
```
11:54:33.377 TypeError: self.validator is not a function
remove_row() run.js:503
jQuery.event.dispatch() jquery.js:4436
jQuery.event.add/elemData.handle() jquery.js:4123
1 run.js:503:13
```
for this line.

File: pkg/docker/run.js [hunk 2]
```diff
@@ -21,905 +21,699 @@ define([
...
-                    row_container_input.attr('disabled', true);
-                }
+            var row_host_input = row.find('input[name="host"]');
+            row_host_input.on('keydown', $.proxy(self, "update", "keydown", "volumes"));
+            row_host_input.on('input', $.proxy(self, "update", "input", "volumes"));
+            row_host_input.on('focusout change', $.proxy(self, "update", "changeFocus", "volumes"));
+            var mount_mode_select = row.find("div .mount-mode");
+            mount_mode_select.find('a').on('click', function() {
+                mount_mode_select.find("button span").text($(this).text());
+                self.update("changeOption", "volumes");
```
Comment (dperpeet) [inline pkg/docker/run.js]: while trying to produce an error, I got
```
11:55:25.278 TypeError: self.update is not a function
render/<() run.js:478
jQuery.event.dispatch() jquery.js:4436
jQuery.event.add/elemData.handle() jquery.js:4123
1 run.js:478:17
```
for this line

**Comment by dperpeet** [comment]
When adding duplicate ports/links, only one message got removed on change (not the other affected):
screenshot from 2016-08-09 12-00-19
Also, if the container port was a duplicate, the host port got highlighted as well (even if it wasn't affected).

**Comment by stefwalter** [comment]
> Not sure if this is intentional: when field level errors in the links (e.g. "Duplicate alias") disappear, the box stays large instead of resizing back to its original size as it currently does on master

FWIW, this is how our pattern works. We use 'visibility: hidden'. Discussion on changing that can happen separately.

File: pkg/docker/run.js (hunk 1 above)
Comment (stefwalter) [inline pkg/docker/run.js]: Fixed.

File: pkg/docker/run.js (hunk 2 above)
Comment (stefwalter) [inline pkg/docker/run.js]: Fixed.

**Comment by stefwalter** [comment]
Okay, I've fixed the 'duplicate port errors'. Remember the intent of the dialog pattern is:

 * No dialog relayout while typing (hence hidden error fields have visibility: hidden, rather than display:none)
 * Assume the user has fixed the error. Actual revalidation happens when running the operation again.

**Comment by stefwalter** [comment]
> >    Not sure if this is intentional: when field level errors in the links (e.g. "Duplicate alias") disappear, the box stays large instead of resizing back to its original size as it currently does on master

> That needs to be fixed.

Just confirming that I was wrong about this "That needs to be fixed".

**Comment by dperpeet** [comment]
> Okay, I've fixed the 'duplicate port errors'. Remember the intent of the dialog pattern is:
> 
>     No dialog relayout while typing (hence hidden error fields have visibility: hidden, rather than display:none)
>     Assume the user has fixed the error. Actual revalidation happens when running the operation again.

I agree completely! Nevertheless I feel that the highlighting loses in comparison with current master. This pull request:
screenshot from 2016-08-11 13-36-20
screenshot from 2016-08-11 13-36-28
screenshot from 2016-08-11 13-36-46
The highlighting doesn't highlight just the relevant boxes and the text doesn't specify either. On master I feel it's more descriptive and the correct error highlights are removed on text change.
screenshot from 2016-08-11 13-40-58
screenshot from 2016-08-11 13-40-52
I don't dispute that the next eval should happen only when running next. What's missing in your pr I believe is logically linking other affected rows so their error highlight gets removed on change.

**Comment by dperpeet** [comment]
That said, I don't think this is a blocker. If you think this is good enough, I will merge and work on my react port, where it's easier to tackle such issues.

**Comment by dperpeet** [comment]
Steps to have the `duplicate` errors not clear properly:
- check expose ports
- add a second port
- set both container ports to the same value, e.g. 80
- click run (will show duplicate error)
- in one row, type something into the host port field
This will clear that one row, but not the duplicate row.

**Comment by stefwalter** [comment]
As discussed on IRC, the clearing of the errors is more forgiving than before. We discussed the various cases and thought that they matched the intent. The intent is that as the user touches things, assume they've fixed the problem ... until the next validation which happens when they click the 'Run' button again.

**Comment by dperpeet** [comment]
Once tests pass, this looks good to be merged. Detailed behavior can be reassessed in the react port.

**Comment by dperpeet** [comment]
tackling some of the races found here in #4869

**Pull request closed without merging**